macid.py

```diff
@@ -0,0 +1,1314 @@
+    def copy(self):
+        model_copy = MACID(node_types=self.node_types)
+        model_copy.add_nodes_from(self.nodes())
+        model_copy.add_edges_from(self.edges())
+        if self.cpds:
+            model_copy.add_cpds(*[cpd.copy() for cpd in self.cpds])
+        return model_copy
+
+
+    def add_cpds(self, *cpds):
+        # this method adds conditional probability distributions to the MACID.
+        for cpd in cpds:
+            if not isinstance(cpd, (TabularCPD, ContinuousFactor, NullCPD)):
+                raise ValueError("Only TabularCPD, ContinuousFactor, or NullCPD can be added.")
+
+            if set(cpd.scope()) - set(cpd.scope()).intersection(set(self.nodes())):
+                raise ValueError("CPD defined on variable not in the model", cpd)
+
+            for prev_cpd_index in range(len(self.cpds)):
+                if self.cpds[prev_cpd_index].variable == cpd.variable:
+                    logging.warning(
+                        "Replacing existing CPD for {var}".format(var=cpd.variable)
+                    )
+                    self.cpds[prev_cpd_index] = cpd
+                    break
+            else:
+                self.cpds.append(cpd)
+
+
+    def check_model(self, allow_null=True):
+        """
+        Check the model for various errors. This method checks for the following
+        errors.
+
+        * Checks if the sum of the probabilities for each state is equal to 1 (tol=0.01).
+        * Checks if the CPDs associated with nodes are consistent with their parents.
+
+        Returns
+        -------
+        check: boolean
+            True if all the checks are passed
+        """
+        for node in self.nodes():
+            cpd = self.get_cpds(node=node)
+
+            if cpd is None:
+                raise ValueError("No CPD associated with {}".format(node))
+            elif isinstance(cpd, (TabularCPD, ContinuousFactor)):
+                evidence = cpd.get_evidence()
+                parents = self.get_parents(node)
+                if set(evidence if evidence else []) != set(parents if parents else []):
+                    raise ValueError(
+                        "CPD associated with {node} doesn't have "
+                        "proper parents associated with it.".format(node=node)
+                    )
+                if not cpd.is_valid_cpd():
+                    raise ValueError(
+                        "Sum or integral of conditional probabilites for node {node}"
+                        " is not equal to 1.".format(node=node)
+                    )
+            elif isinstance(cpd, (NullCPD)):
+                if not allow_null:
+                    raise ValueError(
+                        "CPD associated with {node} is nullcpd".format(node=node)
+                    )
+        return True
+
+
+# --------------   methods for plotting MACID ----------------
+
+    def _get_color(self, node: str):
+    # This colour codes the decision, chance and utility nodes of each agent
```

Comment:
Would be worth commenting that this uses the CID coloring scheme. Add a TODO: use shape to indicate node type, and color to specify the agent. (this can be done later)

Reply:
Good point! Thanks!

Reply:
It now gives changes the shape and colour of the nodes 

In single-agent case: I've kept with the blue rectangle/yellow diamond convention

In the multi-agent case: the colour denotes the agent, the shape denotes the type